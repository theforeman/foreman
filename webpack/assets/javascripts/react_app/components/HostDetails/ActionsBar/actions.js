import React from 'react';
import { FormattedMessage } from 'react-intl';
import { visit } from '../../../../foreman_navigation';
import { foremanUrl } from '../../../common/helpers';
import { sprintf, translate as __ } from '../../../common/I18n';
import { openConfirmModal } from '../../ConfirmModal';
import { APIActions } from '../../../redux/API';
import { HOST_DETAILS_KEY } from '../consts';
import { selectAPIResponse } from '../../../redux/API/APISelectors';
import { POWER_REQURST_KEY } from '../DetailsCard/PowerStatus/constants';

export const deleteHost = (
  hostName,
  compute,
  destroyVmOnHostDelete
) => dispatch => {
  const successToast = () =>
    sprintf(__('Host %s has been removed successfully'), hostName);
  const errorToast = ({ message }) => message;
  const url = foremanUrl(`/api/hosts/${hostName}`);

  // TODO: Replace with a checkbox instead of a global setting for cascade host destroy
  const cascadeMessage = () => {
    if (compute) {
      return destroyVmOnHostDelete
        ? __(
            'This will delete the VM and its disks. This behavior can be changed via global setting "Destroy associated VM on host delete".'
          )
        : __(
            'VM and its disks will not be deleted. This behavior can be changed via global setting "Destroy associated VM on host delete".'
          );
    }
    return null;
  };

  dispatch(
    openConfirmModal({
      isWarning: true,
      title: __('Delete host?'),
      confirmButtonText: __('Delete host'),
      onConfirm: () =>
        dispatch(
          APIActions.delete({
            url,
            key: `${hostName}-DELETE`,
            successToast,
            errorToast,
            handleSuccess: () => visit(foremanUrl('/hosts')),
          })
        ),
      message: (
        <FormattedMessage
          id="delete-host"
          values={{
            host: <b>{hostName}</b>,
            cascade: cascadeMessage(),
          }}
          defaultMessage={__(
            'Are you sure you want to delete host {host}? This action is irreversible. {cascade}'
          )}
        />
      ),
    })
  );
};

export const updateHost = hostId => dispatch => {
  const url = foremanUrl(`/api/hosts/${hostId}`);
  dispatch(
    APIActions.get({
      url,
      key: HOST_DETAILS_KEY,
    })
  );
};

export const buildHost = hostId => dispatch => {
  const successToast = () =>
    sprintf(__('Host %s will be built next boot'), hostId);
  const errorToast = ({ message }) => message;
  const url = foremanUrl(`/hosts/${hostId}/setBuild`);
  dispatch(
    APIActions.put({
      url,
      key: `${hostId}_BUILD`,
      successToast,
      errorToast,
      handleSuccess: () => dispatch(updateHost(hostId)),
    })
  );
};

export const cancelBuild = hostId => dispatch => {
  const successToast = () =>
    sprintf(__('Canceled pending build for %s'), hostId);
  const errorToast = ({ message }) => message;
  const url = foremanUrl(`/hosts/${hostId}/cancelBuild`);
  dispatch(
    APIActions.get({
      url,
      key: `${hostId}_CANCEL_BUILD`,
      successToast,
      errorToast,
      handleSuccess: () => dispatch(updateHost(hostId)),
    })
  );
};

export const isHostTurnOn = store => {
  const { state } = selectAPIResponse(store, POWER_REQURST_KEY);
  return state === 'on';
};
