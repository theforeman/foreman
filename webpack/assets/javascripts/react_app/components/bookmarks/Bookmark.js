import React from 'react';
import PropTypes from 'prop-types';
import URI from 'urijs';
import { MenuItem } from 'patternfly-react';
import EllipisWithTooltip from 'react-ellipsis-with-tooltip';

const Bookmark = ({ text, query }) => {
  const handleClick = e => {
    e.preventDefault();
    const uri = new URI(window.location.href);

    uri.setSearch('search', query.trim());
    window.Turbolinks.visit(uri.toString());
  };

  return (
    <MenuItem onClick={handleClick}>
      <EllipisWithTooltip>{text}</EllipisWithTooltip>
    </MenuItem>
  );
};

Bookmark.propTypes = {
  text: PropTypes.string.isRequired,
  query: PropTypes.string.isRequired,
};

export default Bookmark;
