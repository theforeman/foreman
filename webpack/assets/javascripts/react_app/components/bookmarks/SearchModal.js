import React from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'patternfly-react';

import { noop } from '../../common/helpers';
import BookmarkForm from './form';
import { translate as __ } from '../../../react_app/common/I18n';

const SearchModal = ({ show, onHide, onEnter, title, controller, url }) => (
  <Modal show={show} enforceFocus onHide={onHide} onEnter={onEnter}>
    <Modal.Header closeButton>
      <Modal.Title>{title}</Modal.Title>
    </Modal.Header>
    <Modal.Body>
      <BookmarkForm controller={controller} url={url} onCancel={onHide} />
    </Modal.Body>
  </Modal>
);

SearchModal.propTypes = {
  controller: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  show: PropTypes.bool,
  title: PropTypes.string,
  onHide: PropTypes.func,
  onEnter: PropTypes.func,
};

SearchModal.defaultProps = {
  show: true,
  title: __('Create Bookmark'),
  onHide: noop,
  onEnter: noop,
};

export default SearchModal;
