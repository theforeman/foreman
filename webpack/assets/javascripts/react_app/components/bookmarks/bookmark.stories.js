import React from 'react';
import { storiesOf } from '@storybook/react';
import BookmarkForm from './form';
import BookmarkModal from './SearchModal';
import storeDecorator from '../../../../../stories/storeDecorator';
import Story from '../../../../../stories/components/Story';

storiesOf('Page chunks/Bookmarks', module)
  .addDecorator(storeDecorator)
  .add('Form', () => (
    <Story>
      <BookmarkForm controller="hosts" url="/api/bookmarks" />
    </Story>
  ))
  .add('ModalForm', () => (
    <Story>
      <BookmarkModal controller="hosts" url="/api/bookmarks" />
    </Story>
  ));
