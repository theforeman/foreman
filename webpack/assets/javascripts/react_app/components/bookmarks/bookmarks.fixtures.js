/* eslint-disable camelcase */
import Immutable from 'seamless-immutable';
import { STATUS } from '../../constants';

export const initialState = Immutable({
  bookmarks: {
    showModal: false,
  },
});
export const bookmarks = [
  {
    name: '1111',
    controller: 'hosts',
    query: 'abc',
    public: true,
    id: 52,
    owner_id: 1,
    owner_type: 'User',
  },
  {
    name: '1122',
    controller: 'hosts',
    query: 'abc',
    public: true,
    id: 54,
    owner_id: 1,
    owner_type: 'User',
  },
];

export const afterRequest = Immutable({
  bookmarks: {
    ...initialState,
    hosts: { errors: null, results: [], status: STATUS.PENDING },
  },
});

export const afterSuccess = Immutable({
  bookmarks: {
    ...initialState,
    hosts: { errors: null, results: bookmarks, status: STATUS.RESOLVED },
  },
});

export const afterSuccessNoResults = Immutable({
  bookmarks: {
    ...initialState,
    hosts: { errors: null, results: [], status: STATUS.RESOLVED },
  },
});

export const afterError = Immutable({
  bookmarks: {
    ...initialState,
    hosts: { errors: 'Oops', results: [], status: STATUS.ERROR },
  },
});
