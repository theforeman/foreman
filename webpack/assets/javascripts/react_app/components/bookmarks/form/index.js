import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import { required, length } from 'redux-form-validators';

import { noop } from '../../../common/helpers';
import Form from '../../common/forms/Form';
import TextField from '../../common/forms/TextField';
import * as FormActions from '../../../redux/actions/common/forms';
import { translate as __ } from '../../../../react_app/common/I18n';

const submit = ({ name, query, publik }, dispatch, props) => {
  const { submitForm, url, controller } = props;
  const values = {
    name,
    query,
    public: publik || false,
    controller,
  };

  return submitForm({ url, values, item: 'Bookmark' });
};

const BookmarkForm = ({ handleSubmit, submitting, error, onCancel }) => (
  <Form
    onSubmit={handleSubmit(submit)}
    onCancel={onCancel}
    disabled={submitting}
    submitting={submitting}
    error={error}
  >
    <TextField
      name="name"
      type="text"
      required="true"
      label={__('Name')}
      validate={[required(), length({ max: 254 })]}
    />
    <TextField
      name="query"
      type="textarea"
      required="true"
      label={__('Query')}
      inputClassName="col-md-8"
      validate={[required(), length({ max: 4096 })]}
    />
    <TextField name="publik" type="checkbox" label={__('Public')} />
  </Form>
);

BookmarkForm.propTypes = {
  submitting: PropTypes.bool,
  error: PropTypes.string,
  handleSubmit: PropTypes.func,
  onCancel: PropTypes.func,
};

BookmarkForm.defaultProps = {
  submitting: false,
  error: undefined,
  handleSubmit: noop,
  onCancel: noop,
};

const form = reduxForm({
  form: 'bookmark',
})(BookmarkForm);

export default connect(
  ({ bookmarks }) => ({
    initialValues: { publik: true, query: bookmarks.currentQuery || '' },
  }),
  FormActions
)(form);
