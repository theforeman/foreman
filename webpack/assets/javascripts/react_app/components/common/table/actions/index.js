export { default as getTableItemsAction } from './getTableItemsAction';
export * from './selectionActions';
