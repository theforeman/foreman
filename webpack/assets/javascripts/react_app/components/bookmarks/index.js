import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import EllipisWithTooltip from 'react-ellipsis-with-tooltip';
import { Dropdown, MenuItem, Spinner, Icon } from 'patternfly-react';
import SearchModal from './SearchModal';
import Bookmark from './Bookmark';
import * as BookmarkActions from '../../redux/actions/bookmarks';
import DocumentationUrl from '../common/DocumentationLink';
import { STATUS } from '../../constants';
import { bindMethods, noop } from '../../common/helpers';
import { sprintf, translate as __ } from '../../../react_app/common/I18n';

class BookmarkContainer extends React.Component {
  constructor(props) {
    super(props);
    bindMethods(this, ['handleNewBookmarkClick', 'loadBookmarks']);
  }

  loadBookmarks() {
    const { bookmarks, status, url, controller, getBookmarks } = this.props;

    if (bookmarks.length === 0 && status !== STATUS.PENDING) {
      getBookmarks(url, controller);
    }
  }

  handleNewBookmarkClick() {
    const { showModal, modalClosed, modalOpened, searchQuery } = this.props;

    if (showModal) {
      modalClosed();
    } else {
      modalOpened(searchQuery);
    }
  }

  render() {
    const {
      controller,
      url,
      showModal,
      modalClosed,
      canCreate,
      bookmarks,
      errors,
      status,
      documentationUrl,
    } = this.props;

    return showModal ? (
      <SearchModal
        show={showModal}
        controller={controller}
        url={url}
        onHide={modalClosed}
      />
    ) : (
      <Dropdown pullRight id={controller} onClick={this.loadBookmarks}>
        <Dropdown.Toggle title={__('Bookmarks')}>
          <Icon type="fa" name="bookmark" />
        </Dropdown.Toggle>
        <Dropdown.Menu className="scrollable-dropdown">
          {canCreate && (
            <MenuItem
              key="newBookmark"
              id="newBookmark"
              onClick={this.handleNewBookmarkClick}
            >
              {__('Bookmark this search')}
            </MenuItem>
          )}
          <DocumentationUrl href={documentationUrl} />
          <MenuItem divider />
          <MenuItem header>{__('Saved Bookmarks')}</MenuItem>
          {status === STATUS.PENDING && (
            <li className="loader-root">
              <Spinner size="xs" loading />
            </li>
          )}
          {status === STATUS.RESOLVED &&
            ((bookmarks.length > 0 &&
              bookmarks.map(({ name, query }) => (
                <Bookmark key={name} text={name} query={query} />
              ))) || <MenuItem disabled> {__('None found')}</MenuItem>)}
          {status === STATUS.ERROR && (
            <MenuItem key="bookmarks-errors">
              <EllipisWithTooltip>
                {sprintf('Failed to load bookmarks: %s', errors)}
              </EllipisWithTooltip>
            </MenuItem>
          )}
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}

BookmarkContainer.propTypes = {
  controller: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  searchQuery: PropTypes.string,
  showModal: PropTypes.bool,
  canCreate: PropTypes.bool,
  bookmarks: PropTypes.array,
  errors: PropTypes.string,
  status: PropTypes.string,
  documentationUrl: PropTypes.string,
  modalClosed: PropTypes.func,
  modalOpened: PropTypes.func,
  getBookmarks: PropTypes.func,
};

BookmarkContainer.defaultProps = {
  searchQuery: '',
  showModal: false,
  canCreate: false,
  bookmarks: [],
  errors: '',
  status: null,
  documentationUrl: '',
  modalClosed: noop,
  modalOpened: noop,
  getBookmarks: noop,
};

const mapStateToProps = (
  { bookmarks },
  { data: { controller, searchQuery } }
) => ({
  errors: bookmarks[controller] && bookmarks[controller].errors,
  bookmarks: (bookmarks[controller] && bookmarks[controller].results) || [],
  status: bookmarks[controller] && bookmarks[controller].status,
  showModal: bookmarks.showModal,
  searchQuery,
});

// I'm flatting the props that come from data attribute, this is done to avoid
// having special handling for this.props.data vs this.props in the codebase.
const mergeProps = (stateProps, dispatchProps, { data }) =>
  Object.assign(stateProps, data, dispatchProps);

export default connect(
  mapStateToProps,
  BookmarkActions,
  mergeProps
)(BookmarkContainer);
