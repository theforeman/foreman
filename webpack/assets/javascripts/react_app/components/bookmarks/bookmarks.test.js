import toJson from 'enzyme-to-json';
import { render, mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Dropdown } from 'patternfly-react';
import BookmarksContainer from './';
import API from '../../API';
import {
  initialState,
  afterSuccess,
  afterSuccessNoResults,
  afterRequest,
  afterError,
  bookmarks,
} from './bookmarks.fixtures';
import { onSuccessActions } from '../../redux/actions/bookmarks/bookmarks.fixtures';
import * as BookmarkActions from '../../redux/actions/bookmarks';

jest.mock('../../API');
API.get = jest.fn(() => Promise.resolve({ results: bookmarks }));
const mockStore = configureMockStore([thunk]);

function setup(state = initialState) {
  const props = {
    data: {
      controller: 'hosts',
      url: '/api/bookmarks',
      canCreate: true,
      documentationUrl: '',
    },
  };

  const component = (
    <Provider store={mockStore(state)}>
      <BookmarksContainer {...props} />
    </Provider>
  );

  return {
    props,
    component,
  };
}

describe('bookmarks loading', () => {
  const loadBookmarksScenario = ({ state, getBookmarksCalls }) => {
    jest.mock('../../redux/actions/bookmarks');
    BookmarkActions.getBookmarks = jest
      .fn()
      .mockReturnValue(onSuccessActions[1]);
    mount(setup(state).component)
      .find(Dropdown)
      .simulate('click');
    expect(BookmarkActions.getBookmarks.mock.calls).toHaveLength(
      getBookmarksCalls
    );
    jest.unmock('../../redux/actions/bookmarks');
  };

  const fixtures = {
    'initial state should call getBookmarks': {
      state: initialState,
      getBookmarksCalls: 1,
    },
    'after success state with 0 bookmarks should call getBookmarks': {
      state: afterSuccessNoResults,
      getBookmarksCalls: 1,
    },
    'after error state should call getBookmarks': {
      state: afterError,
      getBookmarksCalls: 1,
    },
    'after request state should not call getBookmarks': {
      state: afterRequest,
      getBookmarksCalls: 0,
    },
    'after success state with bookmarks should not call getBookmarks': {
      state: afterSuccess,
      getBookmarksCalls: 0,
    },
  };

  Object.keys(fixtures).forEach(testCase =>
    it(testCase, () => loadBookmarksScenario(fixtures[testCase]))
  );
});

describe('bookmarks', () => {
  it('empty state', () => {
    expect(toJson(render(setup().component))).toMatchSnapshot();
  });

  it('should show loading spinner when loading bookmarks', () => {
    const spinner = mount(setup(afterRequest).component).find('Spinner');
    expect(spinner).toHaveLength(1);
  });

  it('should show an error message if loading failed', () => {
    expect(toJson(render(setup(afterError).component))).toMatchSnapshot();
  });

  it('should show no bookmarks if server did not respond with any', () => {
    expect(
      toJson(render(setup(afterSuccessNoResults).component))
    ).toMatchSnapshot();
  });
  it('should include existing bookmarks for the current controller', () => {
    expect(mount(setup(afterSuccess).component).find('Bookmark')).toHaveLength(
      2
    );
  });
  it('should not allow creating a new bookmark for users who dont have permission', () => {
    const { props } = setup();

    props.canCreate = false;

    const wrapper = mount(
      <Provider store={mockStore(afterSuccess)}>
        <BookmarksContainer {...props} />
      </Provider>
    );

    expect(wrapper.find('MenuItem#newBookmarks')).toHaveLength(0);
  });
  it('should hide the modal form intiailly', () => {
    expect(
      mount(setup().component)
        .find('BookmarkContainer')
        .props().showModal
    ).toBe(false);
  });

  // eslint-disable-next-line jest/no-disabled-tests
  xit('should open the modal form for new bookmark', () => {
    const wrapper = mount(setup().component);
    expect(wrapper.find('BookmarkContainer').props().showModal).toBe(false);
    wrapper.find('a #newBookmark').simulate('click');
    expect(wrapper.find('BookmarkContainer').props().showModal).toBe(true);
    // TODO: look at alternative solution at https://github.com/airbnb/enzyme/issues/252#issuecomment-266125422
  });

  // eslint-disable-next-line jest/no-disabled-tests
  xit('full flow', () => {
    const wrapper = mount(setup().component);

    expect(wrapper.find('Bookmark')).toHaveLength(0);
    wrapper.find('a #newBookmark').simulate('click');

    const formWrapper = wrapper.find('SearchModal');

    formWrapper
      .find('input [name="name"]')
      .simulate('change', { target: { value: 'Joe.D' } });
    formWrapper
      .find('textarea [name="query"]')
      .simulate('change', { target: { value: 'search' } });
    formWrapper
      .find('input [name="publik"]')
      .simulate('change', { target: { value: false } });

    formWrapper.find('form').simulate('submit');
    expect(wrapper.find('Bookmark')).toHaveLength(1);
  });
});
