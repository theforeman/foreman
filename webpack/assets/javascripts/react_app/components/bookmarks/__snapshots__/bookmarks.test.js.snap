// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookmarks empty state 1`] = `
<Dropdown
  bsClass="dropdown"
  componentClass={[Function]}
  id="hosts"
  onClick={[Function]}
  onToggle={[Function]}
  pullRight={true}
>
  <ButtonGroup
    block={false}
    bsClass="btn-group"
    className="dropdown"
    justified={false}
    onClick={[Function]}
    vertical={false}
  >
    <div
      className="dropdown btn-group"
      onClick={[Function]}
    >
      <DropdownToggle
        bsClass="dropdown-toggle"
        bsRole="toggle"
        id="hosts"
        key=".0"
        onClick={[Function]}
        onKeyDown={[Function]}
        open={false}
        title="Bookmarks"
        useAnchor={false}
      >
        <Button
          active={false}
          aria-expanded={false}
          aria-haspopup={true}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="dropdown-toggle"
          disabled={false}
          id="hosts"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          title="Bookmarks"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle btn btn-default"
            disabled={false}
            id="hosts"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            title="Bookmarks"
            type="button"
          >
            <Icon
              name="bookmark"
              type="fa"
            >
              <FontAwesome
                name="bookmark"
              >
                <span
                  aria-hidden={true}
                  className="fa fa-bookmark"
                />
              </FontAwesome>
            </Icon>
             
            <span
              className="caret"
            />
          </button>
        </Button>
      </DropdownToggle>
      <DropdownMenu
        bsClass="dropdown-menu"
        bsRole="menu"
        className="scrollable-dropdown"
        key=".1"
        labelledBy="hosts"
        onClose={[Function]}
        onSelect={[Function]}
        pullRight={true}
      >
        <RootCloseWrapper
          disabled={true}
          event="click"
          onRootClose={[Function]}
        >
          <ul
            aria-labelledby="hosts"
            className="scrollable-dropdown dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              header={false}
              id="newBookmark"
              key=".$newBookmark"
              onClick={[Function]}
              onKeyDown={[Function]}
              onSelect={[Function]}
            >
              <li
                className=""
                role="presentation"
              >
                <SafeAnchor
                  componentClass="a"
                  id="newBookmark"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="menuitem"
                  tabIndex="-1"
                >
                  <a
                    href="#"
                    id="newBookmark"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1"
                  >
                    Bookmark this search
                  </a>
                </SafeAnchor>
              </li>
            </MenuItem>
            <DocumentationLink
              href=""
              key=".1"
              onKeyDown={[Function]}
              onSelect={[Function]}
            >
              <MenuItem
                bsClass="dropdown"
                disabled={false}
                divider={false}
                header={false}
                href=""
                key="documentationUrl"
                onClick={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    href=""
                    onClick={[Function]}
                    role="menuitem"
                    tabIndex="-1"
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    >
                      <Icon
                        name="question-circle"
                        type="fa"
                      >
                        <FontAwesome
                          name="question-circle"
                        >
                          <span
                            aria-hidden={true}
                            className="fa fa-question-circle"
                          />
                        </FontAwesome>
                      </Icon>
                       Documentation
                    </a>
                  </SafeAnchor>
                </li>
              </MenuItem>
            </DocumentationLink>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={true}
              header={false}
              key=".2"
              onKeyDown={[Function]}
              onSelect={[Function]}
            >
              <li
                className="divider"
                onKeyDown={[Function]}
                role="separator"
              />
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              header={true}
              key=".3"
              onKeyDown={[Function]}
              onSelect={[Function]}
            >
              <li
                className="dropdown-header"
                onKeyDown={[Function]}
                role="heading"
              >
                Saved Bookmarks
              </li>
            </MenuItem>
          </ul>
        </RootCloseWrapper>
      </DropdownMenu>
    </div>
  </ButtonGroup>
</Dropdown>
`;

exports[`bookmarks should show an error message if loading failed 1`] = `
<MenuItem
  bsClass="dropdown"
  disabled={false}
  divider={false}
  header={false}
  key=".$bookmarks-errors"
  onKeyDown={[Function]}
  onSelect={[Function]}
>
  <li
    className=""
    role="presentation"
  >
    <SafeAnchor
      componentClass="a"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        <EllipisWithTooltip>
          <div
            onMouseEnter={[Function]}
            style={
              Object {
                "overflow": "hidden",
                "overflowWrap": "break-word",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
                "wordBreak": "break-all",
              }
            }
          >
            Failed to load bookmarks: %s
          </div>
        </EllipisWithTooltip>
      </a>
    </SafeAnchor>
  </li>
</MenuItem>
`;

exports[`bookmarks should show no bookmarks if server did not respond with any 1`] = `
Array [
  <SafeAnchor
    componentClass="a"
    id="newBookmark"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="menuitem"
    tabIndex="-1"
  >
    <a
      href="#"
      id="newBookmark"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      Bookmark this search
    </a>
  </SafeAnchor>,
  <SafeAnchor
    componentClass="a"
    href=""
    onClick={[Function]}
    role="menuitem"
    tabIndex="-1"
  >
    <a
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      <Icon
        name="question-circle"
        type="fa"
      >
        <FontAwesome
          name="question-circle"
        >
          <span
            aria-hidden={true}
            className="fa fa-question-circle"
          />
        </FontAwesome>
      </Icon>
       Documentation
    </a>
  </SafeAnchor>,
  <SafeAnchor
    componentClass="a"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="menuitem"
    tabIndex="-1"
  >
    <a
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
       
      None found
    </a>
  </SafeAnchor>,
]
`;
